// botframework-webchat-component/lib/Utils/FocusRedirector.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This is an element, when focused, will send the focus to the ref specified in "redirectRef".
// Although the focus is being redirected, browser will scroll this redirector element into view.
// Browser's "scrollIntoView()" call cannot be prevented through event.preventDefault() in both
// bubble and capture phase of "focus" event.
// When this focus redirector is put inside a scrollable container, you may want to resize or reposition
// it to prevent unintentional scroll done by the browser default behavior.
var FocusRedirector = function FocusRedirector(_ref) {
  var className = _ref.className,
      onFocus = _ref.onFocus,
      redirectRef = _ref.redirectRef;
  var handleFocus = (0, _react.useCallback)(function () {
    var current = redirectRef.current;
    current && current.focus();
    onFocus && onFocus();
  }, [onFocus, redirectRef]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    onFocus: handleFocus,
    tabIndex: 0
  });
};

FocusRedirector.defaultProps = {
  className: undefined,
  onFocus: undefined
};
FocusRedirector.propTypes = {
  className: _propTypes.default.string,
  onFocus: _propTypes.default.func,
  redirectRef: _propTypes.default.shape({
    current: _propTypes.default.any
  }).isRequired
};
var _default = FocusRedirector;
exports.default = _default;
