// botframework-webchat-component/lib/Styles/StyleSet/SuggestedAction.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createSuggestedActionStyle;

/* eslint no-magic-numbers: "off" */
function createSuggestedActionStyle(_ref) {
  var accent = _ref.accent,
      paddingRegular = _ref.paddingRegular,
      paddingWide = _ref.paddingWide,
      primaryFont = _ref.primaryFont,
      suggestedActionBackground = _ref.suggestedActionBackground,
      suggestedActionBorderColor = _ref.suggestedActionBorderColor,
      suggestedActionBorderStyle = _ref.suggestedActionBorderStyle,
      suggestedActionBorderWidth = _ref.suggestedActionBorderWidth,
      suggestedActionBorderRadius = _ref.suggestedActionBorderRadius,
      suggestedActionImageHeight = _ref.suggestedActionImageHeight,
      suggestedActionTextColor = _ref.suggestedActionTextColor,
      suggestedActionDisabledBackground = _ref.suggestedActionDisabledBackground,
      suggestedActionDisabledBorderColor = _ref.suggestedActionDisabledBorderColor,
      suggestedActionDisabledBorderStyle = _ref.suggestedActionDisabledBorderStyle,
      suggestedActionDisabledBorderWidth = _ref.suggestedActionDisabledBorderWidth,
      suggestedActionDisabledTextColor = _ref.suggestedActionDisabledTextColor,
      suggestedActionHeight = _ref.suggestedActionHeight,
      suggestedActionsStackedLayoutButtonMaxHeight = _ref.suggestedActionsStackedLayoutButtonMaxHeight,
      subtle = _ref.subtle;
  return {
    '&.webchat__suggested-action': {
      display: 'flex',
      maxWidth: '100%',
      '& .webchat__suggested-action__button': {
        alignItems: 'center',
        borderRadius: suggestedActionBorderRadius,
        fontFamily: primaryFont,
        fontSize: 'inherit',
        height: suggestedActionHeight,
        justifyContent: 'center',
        maxWidth: '100%',
        paddingLeft: paddingWide,
        paddingRight: paddingWide,
        whiteSpace: 'nowrap',
        width: '100%',
        '&:disabled, &[aria-disabled="true"]': {
          background: suggestedActionDisabledBackground || suggestedActionBackground,
          borderColor: suggestedActionDisabledBorderColor,
          borderStyle: suggestedActionDisabledBorderStyle,
          borderWidth: suggestedActionDisabledBorderWidth,
          color: suggestedActionDisabledTextColor || subtle
        },
        '&:not(:disabled):not([aria-disabled="true"])': {
          background: suggestedActionBackground,
          borderColor: suggestedActionBorderColor || accent,
          borderStyle: suggestedActionBorderStyle,
          borderWidth: suggestedActionBorderWidth,
          color: suggestedActionTextColor || accent
        }
      },
      '& .webchat__suggested-action__image': {
        height: suggestedActionImageHeight
      },
      '& .webchat__suggested-action--wrapping': {
        height: 'auto',
        maxHeight: suggestedActionsStackedLayoutButtonMaxHeight || '100%',
        minHeight: typeof suggestedActionsStackedLayoutButtonMaxHeight === 'number' ? Math.min(suggestedActionsStackedLayoutButtonMaxHeight, suggestedActionHeight) : suggestedActionHeight
      },
      '&:not(.webchat__suggested-action--rtl) .webchat__suggested-action__image + .webchat__suggested-action__text': {
        paddingLeft: paddingRegular
      },
      '&.webchat__suggested-action--rtl .webchat__suggested-action__image + .webchat__suggested-action__text': {
        paddingRight: paddingRegular
      }
    }
  };
}
