// botframework-webchat/lib/renderMarkdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = render;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _markdownItForInline = _interopRequireDefault(require("markdown-it-for-inline"));

var _markdownIt = _interopRequireDefault(require("markdown-it"));

var _markdownItAttrsEs = _interopRequireDefault(require("markdown-it-attrs-es5"));

var _sanitizeHtml = _interopRequireDefault(require("sanitize-html"));

/* eslint no-magic-numbers: ["error", { "ignore": [0, 1, 2] }] */
var SANITIZE_HTML_OPTIONS = {
  allowedAttributes: {
    a: ['aria-label', 'href', 'name', 'rel', 'target', 'title'],
    img: ['alt', 'class', 'src']
  },
  allowedSchemes: ['data', 'http', 'https', 'ftp', 'mailto', 'sip', 'tel'],
  allowedTags: ['a', 'b', 'blockquote', 'br', 'caption', 'code', 'del', 'div', 'em', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hr', 'i', 'img', 'ins', 'li', 'nl', 'ol', 'p', 'pre', 's', 'span', 'strike', 'strong', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'ul']
}; // Put a transparent pixel instead of the "open in new window" icon, so developers can easily modify the icon in CSS.

var TRANSPARENT_GIF = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7'; // This is used for parsing Markdown for external links.

var internalMarkdownIt = new _markdownIt.default();

function render(markdown, _ref) {
  var markdownRespectCRLF = _ref.markdownRespectCRLF;

  var _ref2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
      _ref2$externalLinkAlt = _ref2.externalLinkAlt,
      externalLinkAlt = _ref2$externalLinkAlt === void 0 ? '' : _ref2$externalLinkAlt;

  if (markdownRespectCRLF) {
    markdown = markdown.replace(/\n\r|\r\n/g, function (carriageReturn) {
      return carriageReturn === '\n\r' ? '\r\n' : '\n\r';
    });
  }

  var html = new _markdownIt.default({
    breaks: false,
    html: false,
    linkify: true,
    typographer: true,
    xhtmlOut: true
  }).use(_markdownItAttrsEs.default).use(_markdownItForInline.default, 'url_new_win', 'link_open', function (tokens, index) {
    var token = tokens[index];
    token.attrSet('rel', 'noopener noreferrer');
    token.attrSet('target', '_blank');
    var linkOpenToken = tokens.find(function (_ref3) {
      var type = _ref3.type;
      return type === 'link_open';
    });

    var _linkOpenToken$attrs$ = linkOpenToken.attrs.find(function (_ref4) {
      var _ref5 = (0, _slicedToArray2.default)(_ref4, 1),
          name = _ref5[0];

      return name === 'href';
    }),
        _linkOpenToken$attrs$2 = (0, _slicedToArray2.default)(_linkOpenToken$attrs$, 2),
        href = _linkOpenToken$attrs$2[1]; // Adds a new icon if the link is http: or https:.
    // Don't add if it's a phone number, etc.


    if (/^http[s\u017F]?:/i.test(href)) {
      externalLinkAlt && token.attrSet('title', externalLinkAlt);
      var iconTokens = internalMarkdownIt.parseInline("![".concat(externalLinkAlt, "](").concat(TRANSPARENT_GIF, ")"))[0].children;
      iconTokens[0].attrJoin('class', 'webchat__markdown__external-link-icon');
      tokens.splice.apply(tokens, [index + 2, 0].concat((0, _toConsumableArray2.default)(iconTokens)));
    }
  }).render(markdown);
  return (0, _sanitizeHtml.default)(html, SANITIZE_HTML_OPTIONS);
}
